test/test_apps/common/test_app_base.h

```diff
@@ -0,0 +1,1094 @@
+        VkSwapchainKHR                old_swapchain        = VK_NULL_HANDLE;
+        VkAllocationCallbacks*        allocation_callbacks = VK_NULL_HANDLE;
+    } info;
+};
+
+SDL_Window*  create_window_sdl(const char* window_name, bool resizable, int width, int height);
+void         destroy_window_sdl(SDL_Window* window);
+VkSurfaceKHR create_surface_sdl(VkInstance instance, SDL_Window* window, VkAllocationCallbacks* allocator = nullptr);
+void         create_swapchain(SwapchainBuilder& swapchain_builder, Swapchain& swapchain);
+
+VkCommandPool create_command_pool(vkb::DispatchTable const& disp, uint32_t queue_family_index);
+
+struct Sync
+{
+    std::vector<VkSemaphore> available_semaphores;
+    std::vector<VkSemaphore> finished_semaphore;
+    std::vector<VkFence>     in_flight_fences;
+    std::vector<VkFence>     image_in_flight;
+
+    Sync()  = default;
+    ~Sync() = default;
+
+    Sync(const Sync&)            = delete;
+    Sync& operator=(const Sync&) = delete;
+
+    Sync(Sync&&)            = default;
+    Sync& operator=(Sync&&) = default;
+};
+
+Sync create_sync_objects(Swapchain const& swapchain, vkb::DispatchTable const& disp, const int max_frames_in_flight);
+
+std::vector<char> readFile(const std::string& filename);
+
+VkShaderModule createShaderModule(vkb::DispatchTable const& disp, const std::vector<char>& code);
+
+VkShaderModule readShaderFromFile(vkb::DispatchTable const& disp, const std::string& filename);
+
+#define VERIFY_VK_RESULT(message, result)                            \
+    {                                                                \
+        if (result != VK_SUCCESS)                                    \
+            throw gfxrecon::test::vulkan_exception(message, result); \
+    }
+
+struct InitInfo
+{
+    SDL_Window*                window;
+    Instance                   instance;
+    vkb::InstanceDispatchTable inst_disp;
+    VkSurfaceKHR               surface;
+    PhysicalDevice             physical_device;
+    Device                     device;
+    vkb::DispatchTable         disp;
+    Swapchain                  swapchain;
+    std::vector<VkImage>       swapchain_images;
+    std::vector<VkImageView>   swapchain_image_views;
+};
+
+InitInfo device_initialization(const std::string& window_name);
+
+void cleanup_init(InitInfo& init);
+
+void recreate_init_swapchain(InitInfo& init, bool wait_for_idle = true);
+
+class TestAppBase
+{
+  public:
+    void run(const std::string& window_name);
+
+  protected:
+    TestAppBase()                              = default;
+    ~TestAppBase()                             = default;
+    TestAppBase(const TestAppBase&)            = delete;
+    TestAppBase& operator=(const TestAppBase&) = delete;
+    TestAppBase(TestAppBase&&)                 = delete;
+    TestAppBase& operator=(TestAppBase&&)      = delete;
+
+    void recreate_swapchain(bool wait_for_idle);
+
+    virtual void setup();
+    virtual bool frame(const int frame_num) = 0;
```

Comment:
No need for `const` to primitive parameters, in general, as those are copied when functions are called.

Reply:
There's a tiny bit of value here in that the implementation of this function will give an error if it changes frame_num.  So although a prototype is (among other things) a contract with the caller, in this case it's also a bit of a helpful restriction on the implementor.  I dunno if that was the intention.

Reply:
Just noticed the `bool` return type is used by the `test_app_base` to determine whether to exit.
I believe it would be nice to add this info to the function documentation, e.g.:
```c
/// @return `true` to continue running, `false` to exit
```